# [실습4] 데이터 요약하기

## 데이터 불러오기

지난 실습에서 데이터 변환을 통해 "요일", "날씨" 컬럼을 알아보기 쉽게 변환하였습니다. 지난 실습을 통해 정제된 데이터를 불러옵니다.

In [2]:
import numpy as np
import pandas as pd

df=pd.read_csv("./data/seoul_park03.csv")
mm=pd.read_csv("./data/misemunji.csv")

df.head()

날짜 공휴일     날씨  유료합계    어른    청소년  어린이  외국인   단체  무료합계    총계     연  \
0  2016-01-01   O  구름 조금  3359  2799  141.0  419   47    0  1023  4382  2016   
1  2016-01-02   O  구름 많음  5173  4370  203.0  600  100  111  2092  7265  2016   
2  2016-01-03   O  구름 많음  3008  2571  128.0  309   91    0  1549  4557  2016   
3  2016-01-04   X  구름 많음   890   602    NaN  235   51  223   800  1690  2016   
4  2016-01-05   X  구름 많음   416   319   35.0   62   43   47   840  1256  2016   

   월  일 요일  
0  1  1  금  
1  1  2  토  
2  1  3  일  
3  1  4  월  
4  1  5  화

## 데이터 통계값 확인: mean(), min(), max(), median()

데이터의 통계값을 확인하기 위해 각종 집계함수(`mean()`, `min()`, `max()`, `median()` 등)를 사용할 수 있습니다. 예를들어 `mean()`을 활용해 "어른" 컬럼의 평균값을 구하려면 다음과 같이 사용할 수 있습니다.

In [3]:
df["어른"].mean()

3568.5073664825045

전체 일차의 어른 입장객 수의 평균을 구할 수 있습니다. 이번엔 총 입장객 수("총계")가 가장 적은 날의 입장객 수를 알아보도록 하겠습니다.

In [4]:
df["총계"].min()

2

총 인원이 2명만 입장한 날이 있다는 사실을 알 수 있습니다. 여담이지만 해당 날짜인 2017년 3월 28일은 조류독감으로 인해 폐쇄되었던 서울대공원의 재개장을 알리는 언론보도가 나온 날짜로 재개장을 위한 입장 및 집계 시스템의 내부 테스트 정도로 유추해볼 수 있습니다.

### [TODO] 집계함수를 사용해 미세먼지가 가장 높은 날의 수치를 변수 `badair`에 저장합니다.
* "미세먼지" 컬럼 값이 가장 높은 날의 수치를 저장합니다.
* 최대값을 구하는 집계함수는 `max()` 입니다.

In [5]:
# 변수 mm에는 미세먼지 데이터가 저장되어 있습니다.
mm.head()

날짜  미세먼지  초미세먼지     오존  이산화질소  일산화탄소  아황산가스
0  2016-01-01    68     45  0.006  0.047    1.0  0.006
1  2016-01-02    51     31  0.009  0.044    0.7  0.006
2  2016-01-03    96     58  0.013  0.049    0.9  0.007
3  2016-01-04    74     47  0.025  0.025    0.6  0.006
4  2016-01-05    32     16  0.016  0.025    0.4  0.006

In [8]:
# None을 지우고 알맞은 코드를 입력하세요.
badair= mm["미세먼지"].max()
badair

198

## 데이터 전체 통계: describe()

위에서 배운 집계함수를 활용하면 내가 원하는 통계값을 확인할 수 있지만, 때로는 전체 통계값을 보고 이를 통해 데이터의 분석 방향을 결정하기도 합니다. 이를 위해 데이터프레임의 다양한 통계값을 정리해서 보여주는 `describe()`를 사용합니다.

`describe()`는 숫자형 데이터들로 이루어진 컬럼들의 데이터의 갯수(count), 평균(mean), 표준편차(std), 최소값(min), 사분위수(25/50/75%), 최대값(max)를 보여줍니다.

In [9]:
df.describe()

유료합계            어른           청소년           어린이          외국인  \
count   1086.000000   1086.000000   1081.000000   1086.000000  1086.000000   
mean    4595.651934   3568.507366    463.038853    550.260589    75.399632   
std     6461.199603   5456.322438   1424.486882   1013.065306   103.030042   
min        0.000000      0.000000      0.000000      0.000000     0.000000   
25%      606.750000    462.500000     24.000000     43.000000    19.000000   
50%     1697.000000   1282.000000     80.000000    150.000000    43.000000   
75%     5130.500000   3390.750000    298.000000    604.000000    87.000000   
max    40219.000000  38511.000000  17345.000000  20277.000000  1174.000000   

                 단체          무료합계            총계            연            월  \
count   1086.000000   1086.000000   1086.000000  1086.000000  1086.000000   
mean     892.831492   2090.390424   6686.042357  2017.177716     6.438306   
std     2155.583046   2299.475934   8469.956231     0.980313     3.415790   
min        0.000000      2.000000      2.000000  2016.000000     1.000000   
25%        0.000000    667.500000   1322.250000  2016.000000     3.000000   
50%       73.000000   1182.000000   3030.000000  2017.000000     6.000000   
75%      691.500000   2867.750000   8146.500000  2018.000000     9.000000   
max    18420.000000  22789.000000  58688.000000  2019.000000    12.000000   

                 일  
count  1086.000000  
mean     15.670350  
std       8.817452  
min       1.000000  
25%       8.000000  
50%      16.000000  
75%      23.000000  
max      31.000000

`describe()`를 사용해 우리는 입장객의 수가 가장 많았던 날은 58,688명이었다는 것("총계" 컬럼의 max값), 평균적으로 어른, 청소년, 어린이, 외국인 순으로 입장객의 수가 많았다는 것(각 컬럼의 mean값 비교) 등 다양한 정보를 얻을 수 있습니다.

## 데이터 그룹화: groupby()

다음은 데이터를 그룹으로 묶어서 확인할 수 있는 `groupby()`입니다. `groupby()`를 사용하면 특정 기준에 따라 데이터를 정리해서 분석할 수 있습니다. 
예를 들어 서울대공원 데이터를 통해서 날씨에 따른 입장객의 수가 궁금할 때, 우리는 `groupby()`를 사용해서 "날씨"를 기준으로 "총계"값의 평균을 구할 수 있습니다. 

In [11]:
df.groupby("날씨")["총계"].mean()

날씨
구름 많음    6234.844765
구름 조금    7409.122881
눈/비      3038.603604
맑음       7756.225225
흐림       6056.450000
Name: 총계, dtype: float64

예상대로 **날씨가 맑음일 때 가장 평균 입장객 수가 많고, 눈/비일 때 가장 적은 것**을 확인할 수 있습니다. 더 나아가 **날씨가 맑을때와 눈/비가 올 때의 입장객 수가 수치적으로 2배가 넘게 차이가 난다는 것**은 물론 **구름의 많고적음은 입장객의 수에 큰 영향을 끼치지 않는다는 사실**까지 알 수 있습니다. 

이렇게 `groupby()`를 사용하면 데이터를 그룹별로 분할하고 각 그룹에 대한 통계량을 확인하여 특정 기준에 따른 데이터의 추세를 확인할 수 있습니다.

이 데이터를 분할하는 기준은 한 개 이상이 될 수도 있습니다. 앞서 날씨에 따라 입장객의 총 수를 확인하였는데, 입장객의 수에 큰 영향을 미치는 또 하나의 요인은 바로 공휴일 여부입니다. `groupby()`에서 그룹화 기준이 되는 컬럼을 "날씨"와 "공휴일" 2개의 컬럼으로 설정하면 전체 데이터를 날씨에 따라서 그룹으로 한번 묶고, 각 날씨별 그룹 안에서 다시 공휴일 여부에 따라 두 분류로 데이터를 분류합니다.

In [12]:
df.groupby(["날씨","공휴일"])["총계"].mean()

날씨     공휴일
구름 많음  O      12994.262500
       X       3489.903553
구름 조금  O      14797.581081
       X       4034.148148
눈/비    O       4239.027778
       X       2462.400000
맑음     O      14658.260274
       X       4374.691275
흐림     O       9374.957447
       X       3113.622642
Name: 총계, dtype: float64

날씨에 따라 묶었던 데이터를 공휴일 여부에 따라서 데이터를 한번 더 분류했더니 좀 더 자세한 정보들을 알 수 있습니다. **공휴일이 아닌 날의 경우 날씨에 따라 입장객의 수 차이가 그리 크지 않다는 사실**을 알 수 있습니다. 또한 **날씨가 맑음이거나 구름일 때에는 공휴일 여부에 입장객의 수가 따라 월등히 차이가 나지만 눈/비 일때는 오히려 날씨에 따른 입장객 수의 차이가 크지 않음**을 알 수 있습니다.

`groupby()`를 활용해 2개 이상의 컬럼을 확인하는 것 역시 가능합니다. 이 경우 집계 결과가 데이터프레임으로 출력됩니다.


In [13]:
df.groupby(["날씨","공휴일"])[["어른", "어린이"]].mean()

어른          어린이
날씨    공휴일                          
구름 많음 O    8279.325000  1152.375000
      X    1366.106599   244.639594
구름 조금 O    9388.608108  1254.202703
      X    1556.111111   283.037037
눈/비   O    2472.111111   358.277778
      X     903.000000   451.253333
맑음    O    9243.287671  1141.273973
      X    1507.610738   298.335570
흐림    O    5857.063830   779.212766
      X    1320.698113   331.981132

### [TODO] 목요일이면서 공휴일인 날짜의 어른 입장객 수 최대값을 변수 `thursday`에 저장하세요.
* `groupby()`를 사용해 "요일" 컬럼과 "공휴일" 컬럼으로 데이터를 묶을 수 있습니다.
* 최대값을 구하기 위한 집계함수는 `max()` 입니다.

In [30]:
# groupby를 활용하여 "요일", "공휴일" 컬럼으로 데이터를 묶고 "어른"입장객 수 최대값을 확인합니다.
series = df.groupby(["요일", "공휴일"])["어른"].max()

In [31]:
# None을 지우고 공휴일인 목요일의 최대 어른 입장객 수를 저장하세요. 
series.loc['목']['O']

38511

In [32]:
df.groupby(["요일", "공휴일"])["어른"].max().reset_index()

요일 공휴일     어른
0   금   O  25824
1   금   X  10467
2   목   O  38511
3   목   X  12233
4   수   O  19887
5   수   X  10322
6   월   O  26358
7   월   X  14566
8   일   O  28943
9   토   O  28592
10  화   O  19727
11  화   X  13436

## 채점
* **[TODO]** 중 수행하지 않은 부분이 없는지 확인하세요.
* 채점을 위해 아래 코드를 실행한 뒤 우측 상단의 제출 버튼을 눌러주세요.
* 코드 수정시 정상적인 채점이 이루어지지 않습니다.

In [12]:
import os
import json

result = {}
result["problem_1"] = int(badair)
result["problem_2"] = thursday

with open("result.json", "w") as f:
    json.dump(result, f)

TypeError: Object of type Series is not JSON serializable